c2rust-analyze/src/util.rs

```diff
@@ -351,3 +352,70 @@ pub fn is_null_const(constant: Constant) -> bool {
 
 pub trait PhantomLifetime<'a> {}
 impl<'a, T: ?Sized> PhantomLifetime<'a> for T {}
+
+/// Determine if `from` can be safely transmuted to `to`,
+/// which is defined as `*(from as *const To)` being a safe operation,
+/// where `from: *const From` and assuming `*from` already was safe.
+///
+/// Note that this is one-way, and is slightly different from [`core::mem::transmute`],
+/// and more similar to [`core::mem::transmute_copy`].
+///
+/// This forms a non-symmetric (one-way) equivalence relation, named `~` below.
+/// Formally, `A ~ B` iff `*a` and `*(a as *const B)` are safe, where `a: *const A`.
+///
+/// However, safe transmutability is difficult to check completely,
+/// so this function only checks a subset of it,
+/// with these formal rules for all types `A`, `B`:
+///
+/// * `A = B => A ~ B`
+/// * `A ~ B => *A ~ *B`
+/// * `A ~ B => [A] ~ B`
+/// * `A ~ B => [A; N] ~ B`, where `const N: usize`
+/// * `A ~ B => [A] ~ [B; N]`, where `const N: usize`
+/// * `A ~ B => [A; N] ~ [B]`, where `const N: usize`
+/// * `uN ~ iN`, `iN ~ uN`, where `N` is an integer width
```

Comment:
These rules let you derive the following: `u8 = u8 => u8 ~ u8 => [u8; 1] ~ [u8] => *[u8; 1] ~ *[u8]`.  This implies that it's safe to do `*(a as *const *mut [u8])` where `a: *const *mut [u8; 1]`.  But even when `*a` is safe, `*(a as *const *mut [u8])` will often be unsafe, since it reads a two-word `*mut [u8]` from a location that might only contain a one-word `*mut [u8; 1]`.

Reply:
I was thinking carefully about that case and thought it wasn't sound, but it did pass `miri`, so I thought it meant it was, but I think I may have miswritten/mistested it slightly.

Reply:
I'm going to remove the `[A; N] ~ [B]` rule, but keep the `[A] ~ [B; N]` one, as this fails:

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aa387753ff96bbdaf9696d4251bb5132)

```rust
unsafe fn f(a: *const *mut [u8]) {
    let _ = *(a as *const *mut [u8; 1]);
}

fn main() {
    let s = &mut [b'a'] as *mut [u8];
    unsafe {
        f(&s);
    }
}
```

but this works:

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7d25cc382885804299a48c414cdf0d8f)

```rust
unsafe fn f(a: *const *mut [u8; 1]) {
    let _ = *(a as *const *mut [u8]);
}

fn main() {
    let s = &mut [b'a'] as *mut [u8; 1];
    unsafe {
        f(&s);
    }
}
```

Reply:
I believe you have the two code blocks (but not playground links or descriptions) switched.

Reply:
You're right; fixed them.

Reply:
Reverted in 720c2fbb841a4cf83460c59db64b18256c280ebc.